pennylane/math/matrix_manipulation.py

```diff
@@ -102,105 +103,134 @@ def expand_matrix(base_matrix, wires, wire_order=None, sparse_format="csr"):
-    wire_indices = [wire_order.index(wire) for wire in wires]
-    subset_wire_order = wire_order[min(wire_indices) : max(wire_indices) + 1]
-    wire_difference = list(set(subset_wire_order) - set(wires))
-    expanded_wires = wires + wire_difference
 
-    # expand matrix if needed
-    if wire_difference:
-        # we use "csr" to speed up the matrix multiplications done during permutation
-        base_matrix = kron(base_matrix, eye(2 ** len(wire_difference), format="coo"), format="csr")
-        base_matrix.eliminate_zeros()
+def _permute_dense_matrix(matrix, wires, wire_order, batch_dim):
+    """Permute the matrix to match the wires given in `wire_order`.
 
-    # permute the matrix to match the `subset_wire_order` values
-    U = _permutation_sparse_matrix(expanded_wires, subset_wire_order)
-    if U is not None:
-        base_matrix = U.T @ base_matrix @ U
-        base_matrix.eliminate_zeros()
+    Args:
+        matrix (np.ndarray): matrix to permute
+        wires (list): wires determining the subspace that base matrix acts on; a base matrix of
+            dimension :math:`2^n` acts on a subspace of :math:`n` wires
+        wire_order (list): global wire order, which has to contain all wire labels in ``wires``,
+            but can also contain additional labels
+        batch_dim (int or None): Batch dimension. If ``None``, batching is ignored.
 
-    # expand the matrix even further if needed
-    mat = base_matrix
-    num_pre_identities = min(wire_indices)
-    if num_pre_identities > 0:
-        # We use "coo" format because it is faster to initialize, and `scipy.sparse.kron` casts
-        # the matrix to this format anyway
-        pre_identity = eye(2**num_pre_identities, format="coo")
-        mat = kron(pre_identity, mat, format=format)
-        mat.eliminate_zeros()
-
-    num_post_identities = len(wire_order) - max(wire_indices) - 1
-    if num_post_identities > 0:
-        post_identity = eye(2**num_post_identities, format="coo")
-        mat = kron(mat, post_identity, format=format)
-        mat.eliminate_zeros()
-
-    return mat.asformat(format=format)
+    Returns:
+        np.ndarray: permuted matrix
+    """
+    if wires == wire_order:
+        return matrix
+
+    # compute the permutations needed to match wire order
+    perm = [wires.index(wire) for wire in wire_order]
+    num_wires = len(wire_order)
+
+    perm += [p + num_wires for p in perm]
+    if batch_dim:
+        perm = [0] + [p + 1 for p in perm]
+
+    # reshape matrix to match wire values e.g. mat[0, 0, 0, 0] = <0000|mat|0000>
+    # with this reshape we can easily swap wires
+    shape = [batch_dim] + [2] * (num_wires * 2) if batch_dim else [2] * (num_wires * 2)
+    matrix = qml.math.reshape(matrix, shape)
+    # transpose matrix
+    matrix = qml.math.transpose(matrix, axes=perm)
+    # reshape back
+    shape = [batch_dim] + [2**num_wires] * 2 if batch_dim else [2**num_wires] * 2
+    return qml.math.reshape(matrix, shape)
```

Comment:
Could you explain how this representation allows for easy wire swapping? Are you re-expressing the matrix in terms of individual projectors or something?

Reply:
As mentioned in the comment, reshaping like this we match each dimension to a specific `bra` or `ket`. For example:
```python
# mat[0, 1, 1, 0] = <01| mat |10>
```

Knowing this we can easily swap wires by swapping the corresponding indices. Bear in mind that when swapping a wire you need to swap both the `bra` and the `ket`.

Reply:
Basically, it's easier to jus reorder the data than perform matrix multiplication. 

We should consider doing something similar for sparse matrices too.

Reply:
I thought about this, but you can't build multidimensional sparse arrays with scipy...

Reply:
Very nice implementation.